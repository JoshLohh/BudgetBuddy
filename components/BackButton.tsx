import React from 'react';
import { Pressable } from 'react-native';
import { Ionicons } from '@expo/vector-icons';
import { useRouter } from 'expo-router';

export default function BackButton({ color = 'black', size = 24, style = {} }) {
  const router = useRouter();
  return (
    <Pressable onPress={() => router.back()} style={[{ padding: 10 }, style]}>
      <Ionicons name="arrow-back" size={size} color={color} />
    </Pressable>
  );
}
