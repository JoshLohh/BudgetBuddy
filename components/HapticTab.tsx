import { Platform } from 'react-native';
import { BottomTabBarButtonProps } from '@react-navigation/bottom-tabs';
import { PlatformPressable } from '@react-navigation/elements';
import * as Haptics from 'expo-haptics';

export function HapticTab(props: BottomTabBarButtonProps) {
  return (
    <PlatformPressable
      {...props}
      onPressIn={ev => {
        if (Platform.OS === 'ios') {
          Haptics.impactAsync(Haptics.ImpactFeedbackStyle.Light);
        }
        props.onPressIn?.(ev);
      }}
      testID={props.testID}
    />
  );
}
