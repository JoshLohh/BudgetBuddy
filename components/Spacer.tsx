import { View, StyleSheet } from 'react-native';
import React from 'react';

type Dimension = number | `${number}%`;

type SpacerProps = {
  width?: Dimension;
  height?: Dimension;
  testID?: 'spacer';
};

const Spacer = ({ width = '100%', height = 40, testID }: SpacerProps) => {
  return <View style={{ width, height }} testID = {testID} />;
};

export default Spacer;


const styles = StyleSheet.create({})