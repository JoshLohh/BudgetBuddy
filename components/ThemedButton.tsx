import React from 'react';
import { Pressable, StyleSheet, PressableProps, Text } from 'react-native';
import { Colors } from '@/constants/Colors';
import { useColorScheme } from '@/hooks/useColorScheme';

type ThemedButtonProps = PressableProps & {
  lightColor?: string;
  darkColor?: string;
  style?: any;
  children?: React.ReactNode;
  testID?: string;
};

export function ThemedButton({
  style,
  lightColor,
  darkColor,
  children,
  testID,
  ...otherProps
}: ThemedButtonProps) {
  const theme = useColorScheme();

  const backgroundColor =
    theme === 'light' ? lightColor ?? Colors.primary : darkColor ?? Colors.primary;

  return (
    <Pressable
      accessibilityRole="button"
      testID={testID}
      style={({ pressed }) => [
        styles.btn,
        { backgroundColor },
        pressed && styles.pressed,
        style,
      ]}
      {...otherProps}
    >
      {typeof children === 'string' ? <Text>{children}</Text> : children}
    </Pressable>
  );
}

const styles = StyleSheet.create({
  btn: {
    padding: 18,
    borderRadius: 6,
    marginVertical: 10,
  },
  pressed: {
    opacity: 0.6,
  },
});
