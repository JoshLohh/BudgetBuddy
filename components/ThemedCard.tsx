import { useThemeColor } from '@/hooks/useThemeColor';
import { StyleSheet, View , ViewProps, useColorScheme } from 'react-native';
import { Colors } from '../constants/Colors'

const ThemedCard = ({ style, ...props }: ViewProps) => {
    const colorScheme = useColorScheme()
    const theme = Colors[colorScheme === "dark" ? "dark" : "light" ] ?? Colors.light
  
    return (
        <View
            style = {[
                {
                    backgroundColor: theme.uiBackGround,
                    padding: 20,
                    borderRadius: 5,
                },
                style
            ]}
            {...props}
        />
  )

}

export default ThemedCard