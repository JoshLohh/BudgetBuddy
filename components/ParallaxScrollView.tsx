import type { PropsWithChildren, ReactElement } from 'react';
import { StyleSheet, View } from 'react-native';
import Animated, {
  interpolate,
  useAnimatedRef,
  useAnimatedStyle,
  useScrollViewOffset,
} from 'react-native-reanimated';

import { ThemedView } from '@/components/ThemedView';
import { useBottomTabOverflow } from '@/components/ui/TabBarBackground';
import { useColorScheme } from '@/hooks/useColorScheme';

const HEADER_HEIGHT = 250;

type Props = PropsWithChildren<{
  headerImage?: ReactElement;
  headerBackgroundColor?: { dark: string; light: string };
}>;

export default function ParallaxScrollView({
  children = null,
  headerImage,
  headerBackgroundColor = { dark: '#222', light: '#fff' },
}: Props) {
  const colorScheme = useColorScheme() ?? 'light';

  const scrollRef = useAnimatedRef<Animated.ScrollView>();
  const scrollOffset = useScrollViewOffset(scrollRef);

  const bottom = useBottomTabOverflow();

  // Animate the header image for parallax effect
  const headerAnimatedStyle = useAnimatedStyle(() => ({
    transform: [
      {
        translateY: interpolate(
          scrollOffset.value,
          [-HEADER_HEIGHT, 0, HEADER_HEIGHT],
          [-HEADER_HEIGHT / 2, 0, HEADER_HEIGHT * 0.75]
        ),
      },
      {
        scale: interpolate(
          scrollOffset.value,
          [-HEADER_HEIGHT, 0, HEADER_HEIGHT],
          [2, 1, 1]
        ),
      },
    ],
  }));

  return (
    <ThemedView style={styles.container}>
      {/* Header Image */}
      <View style={[styles.header, { backgroundColor: headerBackgroundColor[colorScheme] }]}>
        {headerImage ? (
          <Animated.View style={[StyleSheet.absoluteFill, headerAnimatedStyle]}>
            {headerImage}
          </Animated.View>
        ) : null}
      </View>
      {/* Content */}
      <Animated.ScrollView
        ref={scrollRef}
        contentContainerStyle={[
          styles.content,
          { paddingBottom: bottom + 32 },
        ]}
        showsVerticalScrollIndicator={false}
        bounces
      >
        {children}
      </Animated.ScrollView>
    </ThemedView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'transparent',
  },
  header: {
    height: HEADER_HEIGHT,
    overflow: 'hidden',
    width: '100%',
  },
  content: {
    flexGrow: 1,
    paddingHorizontal: 24,
    paddingTop: 16,
    gap: 16,
    backgroundColor: 'transparent',
  },
});
