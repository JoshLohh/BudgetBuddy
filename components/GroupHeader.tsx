import React, { useState, useEffect } from 'react';
import { View, StyleSheet, ActivityIndicator } from 'react-native';
import { ThemedView } from '@/components/ThemedView';
import { ThemedText } from '@/components/ThemedText';
import ThemedTextInput from '@/components/ThemedTextInput';
import { ThemedButton } from '@/components/ThemedButton';
import { Ionicons } from '@expo/vector-icons';
import { Colors } from '@/constants/Colors';
import { databases } from '@/lib/appwrite';
import { router } from 'expo-router';

const databaseId = process.env.EXPO_PUBLIC_APPWRITE_DATABASE_ID ?? '';
const groupsCollectionId = process.env.EXPO_PUBLIC_APPWRITE_GROUPS_COLLECTION_ID ?? '';

export default function GroupHeader({ group, totalExpenses, onGroupUpdated }) {
  const [editing, setEditing] = useState(false);
  const [title, setTitle] = useState(group?.title || '');
  const [description, setDescription] = useState(group?.description || '');
  const [saving, setSaving] = useState(false);
  const [error, setError] = useState('');

  useEffect(() => {
    setTitle(group?.title || '');
    setDescription(group?.description || '');
  }, [group]);

  if (!group) {
    return (
      <View style={[styles.container, { alignItems: 'center', justifyContent: 'center', minHeight: 80 }]}>
        <ActivityIndicator />
      </View>
    );
  }

  const handleSave = async () => {
    setSaving(true);
    setError('');
    try {
      await databases.updateDocument(databaseId, groupsCollectionId, group.id, {
        title,
        description,
      });
      setEditing(false);
      if (onGroupUpdated) {
        onGroupUpdated({ ...group, title, description });
      }
    } catch (e) {
      setError('Failed to update group details.');
    }
    setSaving(false);
  };

  if (editing) {
    return (
      <ThemedView style={styles.container}>
        <View style={styles.topRow}>
          <View style={styles.leftCol}>
            <ThemedTextInput
              label="Group Name"
              value={title}
              onChangeText={setTitle}
              style={{ marginBottom: 8 }}
            />
            <ThemedTextInput
              label="Description"
              value={description}
              onChangeText={setDescription}
              style={{ marginBottom: 8 }}
            />
            {error ? (
              <ThemedText style={{ color: 'red', marginBottom: 8 }}>{error}</ThemedText>
            ) : null}
          </View>
          <View style={styles.rightCol}>
            <View style={styles.totalCard}>
              <ThemedText style={styles.totalLabel}>Total</ThemedText>
              <ThemedText style={styles.totalAmount}>${totalExpenses.toFixed(2)}</ThemedText>
            </View>
            <ThemedButton
              onPress={handleSave}
              disabled={saving}
              style={styles.editBtn}
            >
              <ThemedText style={{ color: '#fff' }}>
                {saving ? 'Saving...' : 'Save'}
              </ThemedText>
            </ThemedButton>
            <ThemedButton
              onPress={() => setEditing(false)}
              style={[styles.editBtn, { backgroundColor: '#e0e0e0', marginTop: 8 }]}
            >
              <ThemedText style={{ color: '#333' }}>Cancel</ThemedText>
            </ThemedButton>
          </View>
        </View>
      </ThemedView>
    );
  }

  return (
    <ThemedView style={styles.container}>
      <View style={styles.topRow}>
        <View style={styles.leftCol}>
          <ThemedText type="title" style={styles.groupTitle}>{group.title}</ThemedText>
          {group.description ? (
            <ThemedText style={styles.groupDescription}>{group.description}</ThemedText>
          ) : null}
          <ThemedText style={styles.membersCount}>
            {group.members?.length ?? 0} member{(group.members?.length ?? 0) !== 1 ? 's' : ''}
          </ThemedText>
          <ThemedButton
            onPress={() => router.push({ pathname: '/group/[groupId]/history', params: { groupId: group.id } })}
            style={{
              marginTop: 4,           
              marginBottom: 4,        
              height: 32,             
              alignSelf: 'flex-start',    
              justifyContent: 'center',
              alignItems: 'center',
              paddingVertical: 0,
              paddingHorizontal: 14,
              minWidth: 200,
            }}
          >
            <ThemedText style={{ textAlign: 'center', fontSize: 12, color:'#fff' }}>View Activity Log</ThemedText>
          </ThemedButton>
        </View>
        <View style={styles.rightCol}>
          <View style={styles.totalCard}>
            <ThemedText style={styles.totalLabel}>Total</ThemedText>
            <ThemedText style={styles.totalAmount}>${totalExpenses.toFixed(2)}</ThemedText>
          </View>
          <ThemedButton
            onPress={() => setEditing(true)}
            style={styles.editBtn}
          >
            <Ionicons name="pencil" size={18} color="#fff" />
          </ThemedButton>
        </View>
      </View>
    </ThemedView>
  );
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#fff', 
    padding: 10,
    marginBottom: 10,
    borderRadius: 12,
  },
  topRow: {
    flexDirection: 'row',
    alignItems: 'flex-start',
    justifyContent: 'space-between',
  },
  leftCol: {
    flex: 1,
    marginRight: 12,
  },
  rightCol: {
    alignItems: 'flex-end',
    minWidth: 110,
  },
  groupTitle: {
    fontSize: 28,
    fontWeight: '700',
    color: 'black'
  },
  groupDescription: {
    fontSize: 15,
    color: '#555',
    marginBottom: 4,
  },
  membersCount: {
    fontSize: 14,
    color: '#888',
    marginBottom: 8,
  },
  totalCard: {
    backgroundColor: '#b3e0ff', 
    borderRadius: 10,
    paddingVertical: 8,
    paddingHorizontal: 18,
    alignItems: 'center',
    minWidth: 110,
    elevation: 2,
    marginBottom: 10,
    alignSelf: 'flex-end',
  },
  totalLabel: {
    fontSize: 13,
    color: '#1976d2',
    fontWeight: 'bold',
  },
  totalAmount: {
    fontSize: 18,
    color: '#1976d2',
    fontWeight: 'bold',
    marginTop: 2,
  },
  editBtn: {
    backgroundColor: Colors.primary,
    borderRadius: 8,
    paddingVertical: 10,
    paddingHorizontal: 0,
    width: 110,
    alignSelf: 'flex-end',
    alignItems: 'center',
    justifyContent: 'center',
  },
});
