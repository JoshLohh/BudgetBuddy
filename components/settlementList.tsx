import React, { useState } from 'react';
import { View, Alert } from 'react-native';
import { ThemedText } from '@/components/ThemedText';
import { ThemedButton } from '@/components/ThemedButton';
import { Ionicons } from '@expo/vector-icons';
import { Settlement } from '@/types';

interface SettlementListProps {
  suggestedSettlements: Settlement[];
  getUsername: (userId: string) => string;
  settleUp: (from: string, to: string, amount: number) => Promise<void>;
  currentUserId: string;
}

export default function SettlementList({
  suggestedSettlements,
  getUsername,
  settleUp,
  currentUserId,
}: SettlementListProps) {
  
  // Filter settlements for this user
  const userSettlements = suggestedSettlements.filter(
    s => s.from === currentUserId || s.to === currentUserId
  );

  // Calculate net balance
  let netBalance = 0;
  suggestedSettlements.forEach(s => {
    if (s.to === currentUserId) netBalance += s.amount;
    else if (s.from === currentUserId) netBalance -= s.amount;
  });

  // Helper for owed/owing statement and color
  let statement = '';
  let statementColor = '#888';
  if (userSettlements.length === 0) {
    statement = 'No settlements needed.';
  } else if (netBalance > 0) {
    statement = `You are owed $${netBalance.toFixed(2)}`;
    statementColor = '#00cc66';
  } else if (netBalance < 0) {
    statement = `You owe $${Math.abs(netBalance).toFixed(2)}`;
    statementColor = '#ff5050';
  } else {
    statement = 'No settlements needed.';
  }

  const [processing, setProcessing] = useState('');

  const handleSettleUp = (from: string, to: string, amount: number) => {
    Alert.alert(
      'Confirm Settle Up',
      `Are you sure you want to settle $${amount.toFixed(2)} from ${getUsername(from)} to ${getUsername(to)}?`,
      [
        { text: 'Cancel', style: 'cancel' },
        {
          text: 'Confirm',
          onPress: async () => {
            setProcessing(`${from}_${to}_${amount}`);
            await settleUp(from, to, amount);
            setProcessing('');
          },
        },
      ]
    );
  };

  return (
    <View style={{ marginTop: 12 }}>
      <ThemedText type="subtitle" style={{ fontWeight: 'bold', marginBottom: 8 }}>
        Suggested Settlements
      </ThemedText>
      <ThemedText style={{ fontWeight:'bold', color: statementColor }}>
        {statement}
      </ThemedText>
      {userSettlements.length === 0 ? (
        <></>
      ) : (
        userSettlements.map(({ from, to, amount }) => (
          <View
            key={`${from}_${to}_${amount}`}
            style={{ flexDirection: 'row', alignItems: 'center', marginBottom: 8 }}
          >
            <ThemedText style={{ flex: 1 , fontWeight:'bold'}}>
              {getUsername(from)} <Ionicons name="arrow-forward" size={16} color="#1976d2" /> {getUsername(to)}
            </ThemedText>
            <ThemedText style={{ width: 80 , color: '#1e88e5', fontWeight: 'bold'}}>${amount.toFixed(2)}</ThemedText>
            <ThemedButton
              onPress={() => handleSettleUp(from, to, amount)}
              disabled={processing === `${from}_${to}_${amount}`}
              style={{ marginLeft: 10, minWidth: 90 }}
            >
              <ThemedText style={{ color: '#fff' }}>Settle Up</ThemedText>
            </ThemedButton>
          </View>
        ))
      )} 
    </View>
  );
}
