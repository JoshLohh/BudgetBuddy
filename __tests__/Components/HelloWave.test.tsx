import React from 'react';
import { render } from '@testing-library/react-native';
import { HelloWave } from '../../components/HelloWave';
import Animated, { useAnimatedStyle, useSharedValue, withRepeat, withSequence, withTiming } from 'react-native-reanimated';

jest.mock('react-native-reanimated', () => {
  const Reanimated = require('react-native-reanimated/mock');
  // Override specific methods to mock Reanimated 2 behavior
  Reanimated.useSharedValue = jest.fn(() => ({ value: 0 }));
  Reanimated.useAnimatedStyle = jest.fn(() => ({ transform: [{ rotate: '0deg' }] }));
  Reanimated.withRepeat = jest.fn((val, count) => `withRepeat(${val}, ${count})`);
  Reanimated.withSequence = jest.fn((...args) => `withSequence(${args.map(arg => JSON.stringify(arg)).join(', ')})`);
  Reanimated.withTiming = jest.fn((val, options) => `withTiming(${val}, ${JSON.stringify(options)})`);
  return Reanimated;
});

describe('HelloWave', () => {
  it('renders without crashing and shows Hello content', () => {
    const { getByTestId } = render(<HelloWave />);
expect(getByTestId('hello-text')).toBeTruthy();

  });

  it('sets up rotation animation using withRepeat and withSequence', () => {
    const sharedRotation = require('react-native-reanimated').useSharedValue();
    const repeat = require('react-native-reanimated').withRepeat;
    const sequence = require('react-native-reanimated').withSequence;
    const timing = require('react-native-reanimated').withTiming;

    // Render triggers useEffect
    render(<HelloWave />);

    expect(timing).toHaveBeenCalledWith(25, { duration: 150 });
    expect(timing).toHaveBeenCalledWith(0, { duration: 150 });
    expect(sequence).toHaveBeenCalled();
    expect(repeat).toHaveBeenCalledWith(expect.anything(), 4);
  });

  it('applies animated style with rotation', () => {
    const animatedStyle = require('react-native-reanimated').useAnimatedStyle();
    expect(animatedStyle).toEqual({ transform: [{ rotate: '0deg' }] });
  });

  it('matches snapshot', () => {
    const tree = render(<HelloWave />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
