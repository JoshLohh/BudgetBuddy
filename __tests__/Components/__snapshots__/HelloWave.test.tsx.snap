// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelloWave matches snapshot 1`] = `
<View
  style={
    {
      "transform": [
        {
          "rotate": "0deg",
        },
      ],
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#11181C",
        },
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        undefined,
        {
          "fontSize": 28,
          "lineHeight": 32,
          "marginTop": -6,
        },
      ]
    }
    testID="hello-text"
  >
    ðŸ‘‹
  </Text>
</View>
`;
